### DISCUSSION TITLE
Combine with datadog

### DISCUSSION THREAD

**Discussion by KnisterPeter** [awslabs/aws-lambda-web-adapter#506] [General]
Hi there,

we would like to integrate the lambda-adapter with a datadog agent.
According to their docs (https://docs.datadoghq.com/serverless/aws_lambda/installation/nodejs/?tab=containerimage) we have to set the `DD_LAMBDA_HANDLER` env-var to the original handler.
When using the lambda-adapter, there is no real handler, can you give guidance on what to do in this case?

**Comment by bnusunny** [comment]
I'm converting this into a discussion.

**Comment by bnusunny** [comment]
With Lambda Web Adapter, the app runs as a standalone web server in Lambda. So for monitoring, it should be monitored as a web server. Please consult Datadog docs for how to monitor a web server. 

One thing to note for tracing: if you want to capture x-ray traces, LWA forwards the x-ray trace id in the http header 'X-Amzn-Trace-Id', which can be extracted and used for tracing.

**Reply by KnisterPeter** [reply]
Thanks for the quick answer

**Comment by astuyve** [comment]
Hey @KnisterPeter - thanks for opening this discussion. (Thanks for the ping @bnusunny!)

We have some work to do on our end to support this project. Our instrumentation mostly expects the full Lambda payload as a struct in basically each language, so it's not super straightforward.

We're prioritizing this work soon and I can update folks when it's ready.

Thanks!
AJ

**Reply by KnisterPeter** [reply]
That's quite nice. I'd be very interested and already opened a ticket on the datadog site.
So I'm very eager to hear on any progress and be a very happy alpha/beta tester.

**Reply by KnisterPeter** [reply]
@astuyve any progress on this topic? We are eager to test it. 😃

**Reply by b3nk3** [reply]
I too would be interested to know if there has been any progress made on this.

**Reply by KnisterPeter** [reply]
Hi @astuyve, can you give an ETA when this will be ready? I was told that it might be late last year. Any plans on continue with this work or was it droped from schedule?

**Reply by astuyve** [reply]
Hey folks! We're still working on this! I'll keep you updated, thanks!

**Reply by b3nk3** [reply]
hello any update to share @astuyve?

**Reply by alexgallotta** [reply]
We are back working at it!
For now there is a small change in the lwa side on my fork
https://github.com/alexgallotta/aws-lambda-web-adapter/commit/186ddd2f91e541e4de67c298cda7509275fc5b17
Happy to get feed back on that for now. The os set env is not ideal, but the client settings are so deep in the libraries and cannot be changed since it fetched the env variable

The bulk of the change will be on the datadog extension side

**Comment by astuyve** [comment]
@bnusunny We'll probably need to proxy the Lambda Runtime API environment variable, but since wrapper scripts only apply to the function process (not other extensions), we couldn't do it automatically.

Would you merge a PR which honors an `ALTERNATE_LAMBDA_RUNTIME_API` env variable?

**Reply by bnusunny** [reply]
Yes, happy to.

**Comment by alexgallotta** [comment]
This PR has the change to support Datadog
https://github.com/awslabs/aws-lambda-web-adapter/pull/588

**Comment by alexgallotta** [comment]
The support for Datadog is going to be released soon.
If anyone wants to try it before the release, let me know

**Reply by KnisterPeter** [reply]
Yes please. I would love to try. 😊

**Reply by alexgallotta** [reply]
The Datadog support has been released
https://github.com/DataDog/datadog-lambda-extension/releases/tag/v77

For now the support is limited to nodejs. Python should come soon (same version, just a matter of config)

Let me know if you are interested in trying! The steps required are:
1- Set [DD_TRACE_PARTIAL_FLUSH_MIN_SPANS](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_TRACE_PARTIAL_FLUSH_MIN_SPANS) to 1
2- Do not trace the readiness endpoint, in node
```js
const tracer = require("dd-trace").init();
tracer.use("http", {
  blocklist: ["/"],
});
```
for `/` default readiness, otherwise the configured readiness endpoint
3- Set the [AWS_LWA_LAMBDA_RUNTIME_API_PROXY](https://github.com/awslabs/aws-lambda-web-adapter?tab=readme-ov-file#intercepting-request-and-response) to a value like `127.0.0.1:9002` to enable the tracing proxy


NB
this is a new feature that does not work with Datadog Extension compatibility mode

**Reply by alexgallotta** [reply]
and of course the lambda web adapter release with support for `AWS_LWA_LAMBDA_RUNTIME_API_PROXY` or a build after this commit https://github.com/awslabs/aws-lambda-web-adapter/commit/9a966444a47435add01721dfe82c997974574eac

**Reply by bnusunny** [reply]
Nice! I will cut a new release this week.

**Reply by bnusunny** [reply]
@KnisterPeter LWA v0.9.1 is out. Please let us know if it works for you.

**Reply by KnisterPeter** [reply]
I'm already working on building a POC. 👍

**Reply by KnisterPeter** [reply]
@alexgallotta What is a meaningful value for `AWS_LWA_LAMBDA_RUNTIME_API_PROXY`? Should it be just a random IP-Port combination or should it be the application running inside the container?

**Reply by bnusunny** [reply]
See this [comment](https://github.com/awslabs/aws-lambda-web-adapter/discussions/506#discussioncomment-12846043). 

@alexgallotta

**Reply by KnisterPeter** [reply]
@bnusunny Thanks, but it's not obvious for me what to set this to given the comment above. Because it's 'like' a value like this '127.0.0.1:9002'. What is this port?

Another question to @alexgallotta: Why blocklist the readiness endpoint if it's '/'? I would assume that's the entrypoint of my application so why to not trace it? Isn't this sufficient as readiness check?

**Reply by KnisterPeter** [reply]
When configured in our container like this:

```
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.9.1 /lambda-adapter /opt/extensions/lambda-adapter
COPY --from=public.ecr.aws/datadog/lambda-extension:77 /opt/. /opt/
```

And with all the usual `DD_` env variables, the only thing I get in our lambda logs is this:

```
{
    "time": "2025-04-29T15:05:18.113Z",
    "type": "platform.extension",
    "record": {
        "name": "datadog-agent",
        "state": "LaunchError",
        "events": [],
        "errorType": "Unknown"
    }
}
{
    "time": "2025-04-29T15:05:18.113Z",
    "type": "platform.initReport",
    "record": {
        "initializationType": "on-demand",
        "phase": "invoke",
        "status": "error",
        "errorType": "Extension.LaunchError",
        "metrics": {
            "durationMs": 1.246
        }
    }
}
{
    "time": "2025-04-29T15:05:18.115Z",
    "type": "platform.runtimeDone",
    "record": {
        "requestId": "d9fb6eea-b1df-44a7-be8c-64453a5caa5d",
        "status": "error",
        "errorType": "Extension.LaunchError",
        "metrics": {
            "durationMs": 0,
            "producedBytes": 0
        }
    }
}
{
    "time": "2025-04-29T15:05:18.115Z",
    "type": "platform.report",
    "record": {
        "requestId": "d9fb6eea-b1df-44a7-be8c-64453a5caa5d",
        "metrics": {
            "durationMs": 3.321,
            "billedDurationMs": 4,
            "memorySizeMB": 512,
            "maxMemoryUsedMB": 3
        },
        "status": "error",
        "errorType": "Extension.LaunchError"
    }
}
```

**Reply by alexgallotta** [reply]
I will try to help with @KnisterPeter questions!

> What is a meaningful value for AWS_LWA_LAMBDA_RUNTIME_API_PROXY? Should it be just a random IP-Port combination or should it be the application running inside the container?
this is to select a socket for the proxy. The IP must be `127.0.0.1` since it's in the local Lambda sandbox. Unless there is some weird corner case where the Web Adapter would run or intercept requests from another Lambda, but that would likely be broken and not supported.
The port can be anything available, I picked 9002 since it's not used by StatsD or Lambda
https://github.com/DataDog/datadog-lambda-extension/blob/main/bottlecap/src/traces/mod.rs#L16-L22

and not used by other extensions or Lambda business logic I was using

> Why blocklist the readiness endpoint if it's '/'? I would assume that's the entrypoint of my application so why to not trace it? Isn't this sufficient as readiness check?

The readiness checkpoint trace has to be ignored by the Datadog tracer. This is to resolve an ambiguity since there are two traces for each "real request".
By "real request", I meant the one causing the invocation and is coming from API GW/function URL/ALB.

This "real request" is received by the AWS Lambda service, which then starts the Lambda extensions init phase.
During this phase, the Lambda Web Adapter will call the readiness probe, causing one or more traces to be generated by the readiness endpoint.

This diagram could help explain the situation. The orange request is the readiness request

![image](https://github.com/user-attachments/assets/5b196493-3cb2-4887-b6e6-63fc17f6ab76)

Regarding the error in your log, it seems that the extension is failing with `Extension.LaunchError`. I would need more logs (from AWS console) and the Dockerfile to understand better

**Reply by alexgallotta** [reply]
would it be a good idea if I open a PR with a couple of examples?

**Reply by KnisterPeter** [reply]
That would be awesome @alexgallotta

**Reply by KnisterPeter** [reply]
The logs is all I have. I can provide the Dockerfile on Monday

**Reply by alexgallotta** [reply]
Here I added some examples with nodejs and python, and nodejs for zip

https://github.com/awslabs/aws-lambda-web-adapter/pull/600

let me know if they work and any advice on anything I could add or remove!

**Reply by KnisterPeter** [reply]
@alexgallotta I have a working lambda now. By that I mean I managed to have a lambda which sends responses to the client with both extensions active. Unfortunatelly no logs are visible in our datadog account given the respective tags/env/service/version.

I think the initial version of our container failed, because it's a scratch container.
@bnusunny @alexgallotta Do you know why a scratch container would not run with the extensions? Is there some shellout to somewhere?

**Reply by KnisterPeter** [reply]
@bnusunny It seems the lambda-adapter is not working in a scratch container:

```
markusw@STAT-HH-2367:~/source/statista/internal-remix-test (docker-lambda)$ docker run --rm -ti --init --entrypoint /opt/extensions/lambda-adapter -e RUST_BACKTRACE=full test:test

thread 'tokio-runtime-worker' panicked at src/lib.rs:231:67:
called `Result::unwrap()` on an `Err` value: hyper_util::client::legacy::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))
stack backtrace:
   0:          0x12bd93c - <unknown>
   1:          0x116cfa3 - <unknown>
   2:          0x12bcc12 - <unknown>
   3:          0x12bd623 - <unknown>
   4:          0x12bc818 - <unknown>
   5:          0x12eefe8 - <unknown>
   6:          0x12eef49 - <unknown>
   7:          0x12ef56c - <unknown>
   8:          0x116bbcf - <unknown>
   9:          0x1171455 - <unknown>
  10:          0x11f9425 - <unknown>
  11:          0x11f793f - <unknown>
  12:          0x12fb61e - <unknown>
  13:          0x12fe560 - <unknown>
  14:          0x12f0dc8 - <unknown>
  15:          0x12f521f - <unknown>
  16:          0x12ef98b - <unknown>
```

Which system dependencies are required? Any shared libraries? And is there an option to create a static built runtime so it can run in a scratch container?

**Reply by bnusunny** [reply]
Lambda Web Adapter can not be used as entry point for the container. You need to install Lambda Runtime Interface Emulator and use it as entry point. 

https://github.com/aws/aws-lambda-runtime-interface-emulator

**Reply by bnusunny** [reply]
Another option is to deploy a Lambda function and test it in the cloud.

**Reply by KnisterPeter** [reply]
> Another option is to deploy a Lambda function and test it in the cloud.

Ah yes, I see my mistake. In the cloud I have a running container. It's using a debian base image. When I use a scratch based container with just my binaries in there it does fail. Could it be that the extension does not work with a scratch container?

**Reply by alexgallotta** [reply]
Sorry I messed up the pr, the good one is this
https://github.com/awslabs/aws-lambda-web-adapter/pull/602

>  Unfortunatelly no logs are visible in our datadog account given the respective tags/env/service/version.

so you can see metrics and traces but no logs? Or anything at all?

**Reply by KnisterPeter** [reply]
@alexgallotta I cannot see anything in datadog. But at least the app started to work. I also see the logs from the app in Cloudwatch and there are the usual tags added.

**Reply by alexgallotta** [reply]
and the [AWS_LWA_LAMBDA_RUNTIME_API_PROXY](https://github.com/awslabs/aws-lambda-web-adapter?tab=readme-ov-file#intercepting-request-and-response) was set, correct?
In that case, the app should not work unless the Datadog extension is running too.
The issue then could be the DD_API_KEY or some tagging issue.

Have you tried to look at the serverless section of the Datadog UI?

**Reply by KnisterPeter** [reply]
@alexgallotta I've found the Lambda in the serverless section. So it is available and configuration is found. But no logs and traces even though from environment variable configuration it should be.

**Reply by KnisterPeter** [reply]
Now I do see logs. Traces are still missing. This was a configruation issue on my side.
I'll test how it is working and if I find new issues, I'll let you know.

Thanks for your help ❤️

**Reply by KnisterPeter** [reply]
@alexgallotta @bnusunny Still in a docker scratch container the extensions don't work. What is required in such a container to have the extensions running? Are the extensions compiled statically or do we need to add a libc or other shared libraries?

**Reply by alexgallotta** [reply]
Great to hear that! So are you also seeing traces in Datadog or not yet?

Regarding the from scratch, Datadog extension requies some libc bindings, so [musl](https://www.musl-libc.org/) or other static bindings are required

**Reply by KnisterPeter** [reply]
I saw a few traces as well. So this should be solved as well now. Thank you.

**Reply by KnisterPeter** [reply]
Would it be possible to create a static binary without libc requirements for the datadog extension?
From security perspective we would like to not depend on any shared libraries.

**Reply by KnisterPeter** [reply]
The `/opt/extension/datadog-agent` is dynamically linked while the `/opt/datadog-agent-go` and `/opt/datadog_wrapper` are static linked binaries. Would it be possibel to change the extension to be statically linked as well?

**Reply by KnisterPeter** [reply]
These are the linked libraries:
```
root@87075a7fe526:/opt/extensions# ldd datadog-agent 
        linux-vdso.so.1 (0x00007ffc0c70b000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f56dc018000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f56dc013000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f56dc00e000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f56db320000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f56dc009000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f56db13f000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f56dc03c000)
```

**Reply by KnisterPeter** [reply]
When trying to run the setup with a static node binary I get this error:
```
node: symbol lookup error: /app/node_modules/.pnpm/@datadog+native-metrics@3.1.1/node_modules/@datadog/native-metrics/prebuilds/linuxglibc-x64/node-napi.node: undefined symbol: napi_module_register
```

**Reply by alexgallotta** [reply]
Currently Datadog doesn't provide the fully statically linked extension.
Is it strictly related to the Lambda Web Adapter use case? If not, it is better to open a specific ticket/request somewhere else more Datadog-specific

**Reply by alexgallotta** [reply]
I am preparing the Datadog customer facing documentation and I was thinking about linking the examples in https://github.com/awslabs/aws-lambda-web-adapter/pull/602

if I can get a review or looks ready to merge, it would be great so I can link them to the docs!

**Reply by bnusunny** [reply]
@alexgallotta I will review it tomorrow.

**Comment by KnisterPeter** [comment]
@alexgallotta It seems that it's not working with the lambda function-url `response_stream` mode.
I've tried it without datadog and that does seem to run fine. As soon as I've enable the respnse_stream mode with datadog enabled, it does fail to send correct responses.

**Reply by KnisterPeter** [reply]
To be more specific: Whenever I set `AWS_LWA_LAMBDA_RUNTIME_API_PROXY` to `127.0.0.1:9002` it does not work anymore.

**Reply by alexgallotta** [reply]
Thanks for reporting that! That use cases wasn't supported yet afaik. I will check in internally if it's on the roadmap and report back here

**Reply by alexgallotta** [reply]
It seems the implementation for streaming might be non trivial and there is not work allocated yet.
Is it a common use case? It might help prioritizing it

**Reply by KnisterPeter** [reply]
For us this is a showstopper. We use [Remix Streaming](https://remix.run/docs/en/main/guides/streaming) which uses content-transfer chunked to implement this feature. I think next.js out-of-band rendering is similar.

**Reply by lym953** [reply]
Streaming Lambda support is now prioritized on Datadog side. I'll work on it next.

**Reply by KnisterPeter** [reply]
@lym953 When you need a test run let me know. ❤️

**Reply by lym953** [reply]
@KnisterPeter Sure, thanks!

**Reply by lym953** [reply]
@KnisterPeter Could you share what's wrong on your side? I did some quick testing and it looks good so far.
Here's my setup:

Dockerfile:
```
FROM public.ecr.aws/lambda/nodejs:22

COPY --from=public.ecr.aws/datadog/lambda-extension:82 /opt/. /opt/
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.9.1 /lambda-adapter /opt/extensions/lambda-adapter

EXPOSE 8080
WORKDIR "/var/task"
ADD src/package.json /var/task/package.json
ADD src/package-lock.json /var/task/package-lock.json
RUN npm install --omit=dev
ADD src/ /var/task
RUN chmod +rx -R /var/task/node_modules
RUN chmod +rx /var/task/index.js
CMD ["index.handler"]
```
Handler:
```
export const handler = awslambda.streamifyResponse(
  async (event, responseStream, context) => {
    responseStream.setContentType("text/plain");

    for (let i = 0; i < 5; i++) {
      console.log(`Writing ${i + 1}`);
      responseStream.write(`Hello, world! (${i + 1})\n`);
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
    responseStream.end();
  }
);
```
Env var:
- `AWS_LWA_INVOKE_MODE`: `response_stream`

Function URL's invoke mode: `RESPONSE_STREAM`

And I see the logs in Datadog:
<img width="657" height="366" alt="image" src="https://github.com/user-attachments/assets/f89b065e-5e00-4548-93bb-69050e5b4e45" />

**Reply by KnisterPeter** [reply]
Hi @lym953 thanks for pointing that out. I'll need to restore our setup (it's a bit dated already) and will report back any errors I see.

**Reply by KnisterPeter** [reply]
Hi @lym953 since the lambda web-adapter should be plain docker container without specifics tied to the lambda API, I'm confused that you use the lambda streaming API. It should work just as an app using content-encoding chunked.

**Reply by lym953** [reply]
@KnisterPeter Thanks for the suggestion. Let me try to test it without Lambda streaming API.

**Reply by lym953** [reply]
I'm using this HTTP handler and still don't see any error. Logs appeared in Datadog as expected.
```
import express from "express";
const app = express();
const port = process.env["PORT"] || 8080;

app.get("/", async (req, res) => {
  res.writeHead(200, {
    "Content-Type": "text/plain",
    "Transfer-Encoding": "chunked",
    "Cache-Control": "no-cache",
    Connection: "keep-alive",
  });

  // Log to stderr to avoid double logging in CloudWatch Logs
  console.error("Starting chunked response...");

  for (let i = 0; i < 5; i++) {
    const chunk = `Hello, world! (${i + 1})\n`;
    console.error(`Sending chunk ${i + 1}: "${chunk.trim()}"`);
    res.write(chunk);
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }

  res.end();
});
```
Could you share your setup and what's wrong when they are available?

**Reply by KnisterPeter** [reply]
@lym953 I try to recreate my setup to see the error. So far it does seem to work, but I can't see any APM traces (just logs). But that could be easily an issue on my side. Give me a few days to inspect this.

**Reply by KnisterPeter** [reply]
Do you see traces in datadog with your setup? The logs are visible on our end as well, but without traces datadog would just too expensive as a pure logsink. 😄

**Reply by KnisterPeter** [reply]
Also in the APM service catalog we don't see the deployment with the lambda-web-adapter in our env.
Only our default lambda setup (without docker) is working as expected.

**Reply by KnisterPeter** [reply]
Okay I can confirm that it does work instead of throwing errors. So streaming is somehow suported, but indeed APM is not working. That limits the usefulness of the feature very much.

**Reply by lym953** [reply]
@KnisterPeter Thanks for testing! Let me see how to support it.

**Reply by KnisterPeter** [reply]
@lym953 any update on this?

**Reply by lym953** [reply]
@KnisterPeter Hey I'm still working on this.

**Reply by lym953** [reply]
@KnisterPeter Could you try extension v84? @duncanista added this support.
Here's an example app: https://github.com/awslabs/aws-lambda-web-adapter/pull/608

**Reply by KnisterPeter** [reply]
Thank you @lym953. I'm going to test that next week at the company.

**Reply by KnisterPeter** [reply]
@lym953 I can report back that it does work now. Thank you for supporting that. ❤️
